In [5]:
import os
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, Normalizer, StandardScaler, MinMaxScaler
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import TomekLinks
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neural_network import MLPRegressor
from imblearn.under_sampling import RandomUnderSampler
from sklearn.tree import DecisionTreeClassifier

In [2]:
os.chdir("C:\\Users\\Seijsener\\Downloads")

In [3]:
churnData = pd.read_csv("Customer-Churn.csv")

In [17]:
churnData.isna().sum()

gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [18]:
churnData.sort_index(inplace=True)

for index, row in churnData.iterrows():
    if pd.isnull(row['TotalCharges']):
        churnData.at[index, 'TotalCharges'] = churnData.at[index - 1, 'TotalCharges'] + row['MonthlyCharges']
print(churnData)

      gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

     OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0                No          Yes               No          No         

In [19]:
churnData.isna().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [27]:
churnData.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

In [60]:
# Perform one-hot encoding for categorical variables
X_encoded = pd.get_dummies(X)

# Display the first few rows of the encoded dataframe
print(X_encoded.head())

   SeniorCitizen  tenure  MonthlyCharges  TotalCharges  gender_Female  \
0              0       1           29.85         29.85           True   
1              0      34           56.95       1889.50          False   
2              0       2           53.85        108.15          False   
3              0      45           42.30       1840.75          False   
4              0       2           70.70        151.65           True   

   gender_Male  Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  ...  \
0        False       False         True           True           False  ...   
1         True        True        False           True           False  ...   
2         True        True        False           True           False  ...   
3         True        True        False           True           False  ...   
4        False        True        False           True           False  ...   

   TechSupport_Yes  StreamingTV_No  StreamingTV_No internet service  \
0            Fa

In [61]:
X = churnData.drop(columns=['Churn'])
y = churnData['Churn']

In [62]:
smote = SMOTE(random_state=42)
X_smote, y_smote = smote.fit_resample(X_encoded, y)

In [63]:
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

In [64]:
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

C:\Users\Seijsener\.conda\New folderX\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [65]:
tree_model = DecisionTreeClassifier()
tree_model.fit(X_train, y_train)

DecisionTreeClassifier()

In [66]:
logistic_pred = logistic_model.predict(X_test)
tree_pred = tree_model.predict(X_test)

In [67]:
logistic_accuracy = accuracy_score(y_test, logistic_pred)
tree_accuracy = accuracy_score(y_test, tree_pred)

print("Logistic Regression Accuracy:", logistic_accuracy)
print("Decision Tree Classifier Accuracy:", tree_accuracy)

Logistic Regression Accuracy: 0.8026969481902059
Decision Tree Classifier Accuracy: 0.730305180979418


Based on these accuracies, the logistic regression model performs better than the decision tree classifier model on this dataset. 

In [68]:
# Apply TomekLinks for downsampling

tomek_links = TomekLinks()
X_resampled, y_resampled = tomek_links.fit_resample(X_encoded, y)

In [69]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

In [70]:
logistic_model = LogisticRegression(random_state=42)
logistic_model.fit(X_train, y_train)


LogisticRegression(random_state=42)

In [71]:
logistic_pred = logistic_model.predict(X_test)
logistic_accuracy = accuracy_score(y_test, logistic_pred)
print("Logistic Regression Accuracy:", logistic_accuracy)

Logistic Regression Accuracy: 0.7939814814814815


In [72]:
tree_model = DecisionTreeClassifier(random_state=42)
tree_model.fit(X_train, y_train)


DecisionTreeClassifier(random_state=42)

In [73]:
tree_pred = tree_model.predict(X_test)
tree_accuracy = accuracy_score(y_test, tree_pred)
print("Decision Tree Classifier Accuracy:", tree_accuracy)

Decision Tree Classifier Accuracy: 0.7361111111111112


In [74]:
# Apply TomekLinks for downsampling
tomek_links = TomekLinks()
X_resampled_tomek, y_resampled_tomek = tomek_links.fit_resample(X_resampled, y_resampled)

# Check the class distribution after TomekLinks
print("Class distribution after TomekLinks:")
print(y_resampled_tomek.value_counts())

# Split the downsampled dataset into training and testing sets
X_train_tomek, X_test_tomek, y_train_tomek, y_test_tomek = train_test_split(X_resampled_tomek, y_resampled_tomek, test_size=0.2, random_state=42)

# Instantiate and train logistic regression model on the TomekLinks downsampled data
logistic_model_tomek = LogisticRegression(random_state=42)
logistic_model_tomek.fit(X_train_tomek, y_train_tomek)

# Predict on the testing set and compute accuracy for logistic regression model with TomekLinks
logistic_pred_tomek = logistic_model_tomek.predict(X_test_tomek)
logistic_accuracy_tomek = accuracy_score(y_test_tomek, logistic_pred_tomek)
print("Logistic Regression Accuracy with TomekLinks:", logistic_accuracy_tomek)

# Instantiate and train decision tree classifier model on the TomekLinks downsampled data
tree_model_tomek = DecisionTreeClassifier(random_state=42)
tree_model_tomek.fit(X_train_tomek, y_train_tomek)

# Predict on the testing set and compute accuracy for decision tree classifier model with TomekLinks
tree_pred_tomek = tree_model_tomek.predict(X_test_tomek)
tree_accuracy_tomek = accuracy_score(y_test_tomek, tree_pred_tomek)
print("Decision Tree Classifier Accuracy with TomekLinks:", tree_accuracy_tomek)


Class distribution after TomekLinks:
Churn
No     4450
Yes    1869
Name: count, dtype: int64
Logistic Regression Accuracy with TomekLinks: 0.8299050632911392
Decision Tree Classifier Accuracy with TomekLinks: 0.754746835443038


C:\Users\Seijsener\.conda\New folderX\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


It seems like after applying TomekLinks, the class distribution has changed. The number of instances in the "Yes" class (churned customers) has decreased, while the number of instances in the "No" class (non-churned customers) has remained relatively stable.
Logistic Regression Accuracy with TomekLinks: 0.8299
Decision Tree Classifier Accuracy with TomekLinks: 0.7547
The logistic regression model's accuracy has increased, while the decision tree classifier's accuracy has slightly decreased compared to the results obtained without TomekLinks.